cdc/entry/schema/snapshot.go

```diff
@@ -0,0 +1,1145 @@
+			zap.Stringer("tableInfo", tableInfo),
+			zap.Bool("ineligible", s.doIsIneligibleTableID(tableInfo.ID)))
+		id, ok := s.doTableIDByName(tableInfo.TableName.Schema, tableInfo.TableName.Table)
+		if !ok || id != tableInfo.ID {
+			logger("[SchemaSnap] ----> tableNameToID item lost", zap.Stringer("name", tableInfo.TableName), zap.Int64("tableNameToID", id))
+		}
+	})
+	s.IterTableNames(func(schemaID int64, table string, target int64) {
+		if _, ok := availableTables[target]; !ok {
+			name := fmt.Sprintf("%s.%s", availableSchemas[schemaID], table)
+			logger("[SchemaSnap] ----> tables item lost", zap.String("name", name), zap.Int64("table", target))
+		}
+	})
+
+	s.IterPartitions(true, func(pid int64, table *model.TableInfo) {
+		logger("[SchemaSnap] --> Partitions", zap.Int64("partitionID", pid), zap.Int64("tableID", table.ID),
+			zap.Bool("ineligible", s.doIsIneligibleTableID(pid)))
+	})
+}
+
+// SchemaByID returns the DBInfo by schema id.
+// The second returned value is false if no schema with the specified id is found.
+// NOTE: The returned table info should always be READ-ONLY!
+func (s *Snapshot) SchemaByID(id int64) (*timodel.DBInfo, bool) {
+	s.rwlock.RLock()
+	defer s.rwlock.RUnlock()
+	return s.doSchemaByID(id)
+}
+
+func (s *Snapshot) doSchemaByID(id int64) (val *timodel.DBInfo, ok bool) {
+	tag := negative(s.currentTs)
+	start := versionedID{id: id, tag: tag, target: nil}
+	end := versionedID{id: id, tag: negative(uint64(0)), target: nil}
+	s.schemas.AscendRange(start, end, func(i btree.Item) bool {
+		val = (*timodel.DBInfo)(i.(versionedID).target)
+		ok = val != nil
+		return false
+	})
+	return
+}
+
+// PhysicalTableByID returns the TableInfo by table id or partition id.
+// The second returned value is false if no table with the specified id is found.
+// NOTE: The returned table info should always be READ-ONLY!
+func (s *Snapshot) PhysicalTableByID(id int64) (*model.TableInfo, bool) {
+	s.rwlock.RLock()
+	defer s.rwlock.RUnlock()
+	return s.doPhysicalTableByID(id)
+}
+
+func (s *Snapshot) doPhysicalTableByID(id int64) (tableInfo *model.TableInfo, ok bool) {
+	tag := negative(s.currentTs)
+	start := versionedID{id: id, tag: tag, target: nil}
+	end := versionedID{id: id, tag: negative(uint64(0)), target: nil}
+	s.tables.AscendRange(start, end, func(i btree.Item) bool {
+		tableInfo = (*model.TableInfo)(i.(versionedID).target)
+		ok = tableInfo != nil
+		return false
+	})
+	if !ok {
+		// Try partition, it could be a partition table.
+		s.partitions.AscendRange(start, end, func(i btree.Item) bool {
+			tableInfo = (*model.TableInfo)(i.(versionedID).target)
+			ok = tableInfo != nil
+			return false
+		})
+	}
+	return
+}
+
+// SchemaIDByName gets the schema id from the given schema name.
+func (s *Snapshot) SchemaIDByName(schema string) (int64, bool) {
+	s.rwlock.RLock()
+	defer s.rwlock.RUnlock()
+	return s.doSchemaIDByName(schema)
+}
+
+func (s *Snapshot) doSchemaIDByName(schema string) (id int64, ok bool) {
```

Comment:
```suggestion
func (s *Snapshot) schemaIDByNameLocked(schema string) (id int64, ok bool) {
```

I think it's more clear when the method should be invoked.